### PR TITLE
Enhance gdal capabilities for complex datasets (squashed)

### PR DESCRIPTION
**Pull request by jmichel-otb**
This pull request creates a new metadata domain, called DERIVED_SUBDATASETS, for all datasets supported by gdal.

This new metadata domain reports derived subdatasets that actually are derived vrt raster band using some of the pixel functions from Valentino. Derived datasets are only exposed if there is at least one raster band, and only derived datasets that make sense are exposed (complex related ones are exposed only if there is at least one complex raster band). The used pixel functions where integrated in another pull request, that has been merged in this one (see #141). All metadata, geotransform and projection reference are imported from the orginal dataset. For the sake of numerical precision, derived bands will have Float64 or CFloat64 precision.

Implementation details:

    - Reading of derived subdatasets are handled by a new dataset class (frtms/derived/deriveddataset.cpp)
    - Reporting the new metadata domain and its entries is done at gdaldataset class level, within GetMetaDataDomainList() and GetMetaData() methods,
    - Mapping between pixel functions and available derived dataset is done in a static array in frmts/dervied/derivedlist.c. Adding new derived datasets should be easy.

Here is an example on Sentinel1 dataset :

$ gdalinfo -nogcp -mdd DERIVED_SUBDATASETS s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
Driver: GTiff/GeoTIFF
Files: s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
Size is 17663, 31106
Coordinate System is `'
Metadata:
AREA_OR_POINT=Area
TIFFTAG_DATETIME=2015:06:20 01:52:30
TIFFTAG_IMAGEDESCRIPTION=Sentinel-1A SM SLC L1
TIFFTAG_SOFTWARE=Sentinel-1 IPF 002.45
Metadata (DERIVED_SUBDATASETS):
DERIVED_SUBDATASET_0_NAME=DERIVED_SUBDATASET:AMPLITUDE:s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_0_DESC=Amplitude of input bands from s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_1_NAME=DERIVED_SUBDATASET:PHASE:s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_1_DESC=Phase of input bands from s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_2_NAME=DERIVED_SUBDATASET:REAL:s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_2_DESC=Real part of input bands from s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_3_NAME=DERIVED_SUBDATASET:IMAG:s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_3_DESC=Imaginary part of input bands from s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_4_NAME=DERIVED_SUBDATASET:CONJ:s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_4_DESC=Conjugate of input bands from s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_5_NAME=DERIVED_SUBDATASET:INTENSITY:s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_5_DESC=Intensity (squared amplitude) of input bands from s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_6_NAME=DERIVED_SUBDATASET:LOGAMPLITUDE:s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
DERIVED_SUBDATASET_6_DESC=log10 of amplitude of input bands from s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
Image Structure Metadata:
INTERLEAVE=BAND
Corner Coordinates:
Upper Left ( 0.0, 0.0)
Lower Left ( 0.0,31106.0)
Upper Right (17663.0, 0.0)
Lower Right (17663.0,31106.0)
Center ( 8831.5,15553.0)
Band 1 Block=17663x1 Type=CInt16, ColorInterp=Gray

And :

$ gdalinfo -nogcp DERIVED_SUBDATASET:AMPLITUDE:s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
Driver: COMPLEXDERIVED/Complex derived bands
Files: s1a-s6-slc-vv-20150619t195043-20150619t195101-006447-00887d-001.tiff
Size is 17663, 31106
Coordinate System is `'
Origin = (0.000000000000000,0.000000000000000)
Pixel Size = (1.000000000000000,1.000000000000000)
Metadata:
AREA_OR_POINT=Area
TIFFTAG_DATETIME=2015:06:20 01:52:30
TIFFTAG_IMAGEDESCRIPTION=Sentinel-1A SM SLC L1
TIFFTAG_SOFTWARE=Sentinel-1 IPF 002.45
Corner Coordinates:
Upper Left ( 0.0000000, 0.0000000)
Lower Left ( 0.000, 31106.000)
Upper Right ( 17663.000, 0.000)
Lower Right ( 17663.000, 31106.000)
Center ( 8831.500, 15553.000)
Band 1 Block=128x128 Type=Float64, ColorInterp=Undefined

### CODE REVIEW

**Comment by rouault** [comment]
Merged in r34651 (with rem some indentation fixes, and a change in numbering of derived_subdatasets so that it starts at 1 like regular subdatasets, and also to avoid it starting at 6 on real datasets) + r34652 for a few Windows compilation fixes

**Comment by rouault** [comment]
All builds are happy. Closing

**Comment by jmichel-otb** [comment]
Thanks a lot for your help and patience. Would not have been able to complete this work without it.

File: gdal/frmts/derived/derivedlist.c [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+static const DerivedDatasetDescription asDDSDesc [] =
+{
+  { "AMPLITUDE", "Amplitude of input bands", "mod", "complex","Float64"},
+  { "PHASE", "Phase of input bands", "phase", "complex","Float64"},
+  { "REAL", "Real part of input bands", "real", "complex","Float64"},
+  { "IMAG", "Imaginary part of input bands", "imag", "complex","Float64"},
+  { "CONJ", "Conjugate of input bands", "conj", "complex","CFloat64"},
+  { "INTENSITY", "Intensity (squared amplitude) of input bands", "intensity", "complex","Float64"},
+  { "LOGAMPLITUDE", "log10 of amplitude of input bands", "log10", "all","Float64"}
```
Comment (avalentino) [inline gdal/frmts/derived/derivedlist.c]: > { "LOGAMPLITUDE", "log10 of amplitude of input bands", "log10", "all","Float64"}

Maybe using the dB pixel function (20+log10(abs(x))) would be even better

**Pull request closed without merging**